### DISCUSSION TITLE
Compose: Bigger Picture

### DISCUSSION THREAD

**Discussion by flatcap** [neomutt/neomutt#3225] [Development]
I have little to write here, as it's still an early WIP.
Here are some technical challenges I think need fixing.

**Comment by flatcap** [comment]
### Eliminate `AttachCtx` and `AttachPtr`

Eliminate `AttachCtx` and `AttachPtr` from all the code except Compose and View Attachments (recv)
These structs contain GUI info and shouldn't be used in attachment library functions.

**Comment by flatcap** [comment]
### Compose: Insert a top-level `Body`

When two attachments are added to an Email, we get the layout:
```
 E        E         E
(B)      (B)B      (B)BB      
```
An `Email` with three `Body`s -- (the bracketed one is the message)
This is `Email->body` (down) and `Body->next` (across)

Just before the `Email` is sent, it has an extra `Body` inserted:
```
 E
 B [multipart/mixed]
(B)BB      
```

If we want to convert `Body` to use `TAILQ` or `ARRAY`, then we need to handle trees of `Body` consistently.
If we insert the `B [multipart/mixed]` in Compose, then all the attachment layouts can be passed to functions as `Body *`.

**Reply by dcpurton** [reply]
For clarity in this thread the `multipart/mixed` attachment is only added if it's required (i.e., when there are multiple attachments at the top level).

**Comment by flatcap** [comment]
### Three Layers of Attachments

We can think of the Attachment API in three layers:
1. Business Logic
2. GUI Components (Views)
3. Data Components (Attachments)

For the case of `<group-multilingual>`, `op_attachment_group_lingual()` would
- Business:
  - Check attachments are at the same level
  - Check attachment types have language info
- GUI:
  - Call `attachment_view` functions to manipulate the `AttachCtx` / `AttachPtr`
- Data:
  - Call `attachment` functions to manipulate the `Body`s

**Comment by flatcap** [comment]
### Eliminate `Body` sibling links

This needs more investigation, but I think it might be possible to eliminate `Body->next`

Very little code actually manipulates the tree of `Body`s.

### Receiving

A received email is parsed giving a `Body`
If you view the list of attachments, more `Body`s are added
If you decrypt the `Email`, then some more `Body`s are added (out-of-tree)

These are simple operations that add to a tree.  No reordering.

### Sending

When you send an email, there are up to three stages:
- A multipart `Body` is added if necessary
- Parts are signed if necessary
- Parts are encrypted if necessary

To create the multipart `Email`, a new `Body` is simply inserted between the `Email` and the existing `Body`s.

To sign an `Email`, two new `Body`s are created.
Again, they are inserted between the `Email` and the existing `Body`s.

To encrypt an `Email`, the `Body`s contents are encrypted.
The existing `Body`s are replaced completely with a new `Body`.

### Composing

When composing an `Email`, the user needs to be able to add, remove and reorder parts at will.
Currently, this is a pain to code because:
- `Body` is singly-linked
- There's a shadow set of objects (`AttachCtx`, `AttachPtr`) to keep in sync

If we add lots of linking to the `AttachPtr`, then we can eliminate `Body->next`, `Body->parts`
and replace them with a simple `ARRAY`.
```c
struct Body
{
  struct BodyArray *children;
};
ARRAY_HEAD(BodyArray, struct Body *);
```

All the existing users of `Body` would be converted/simplified to use `ARRAY_FOREACH()`.
